birdhouse/components/weaver/config/magpie/weaver_hooks.py

```diff
@@ -0,0 +1,112 @@
+    # if explicitly provided, ensure it is permitted (admin allow any, otherwise self-user reference only)
+    if header is not None:
+        if request.user is None:
+            header = "public"
+        else:
+            if not is_admin(request):
+                # override disallowed writing to other location
+                # otherwise, up to admin to have writen something sensible
+                header = "user-" + str(request.user.id)
+    else:
+        if request.user is None:
+            header = "public"
+        else:
+            header = "user-" + str(request.user.id)
+    request.headers["X-WPS-Output-Context"] = header
+    return request
+
+
+def filter_allowed_processes(response, context):
+    # type: (Response, HookContext) -> Response
+    """
+    Filter processes returned by Weaver response according to allowed resources by user.
+    """
+    if "application/json" in response.content_type:
+        body = response.json
+        if "processes" in body:
+            if is_admin(response.request):  # don't waste time checking permissions, full access anyway
+                return response
+
+            # depending on 'detail' query, processes can be returned as list of IDs or nested JSON summaries
+            processes = {
+                proc if isinstance(proc, str) else proc.get("id"): proc
+                for proc in body["processes"]
+            }
+
+            # only need 2 first levels ('processes' and each process 'id' under it)
+            children = ru.get_resource_children(context.resource, response.request.db, limit_depth=2)
+            proc_res = None
+            for res in children.values():
+                if res["node"].resource_name == "processes":
+                    # if nothing under 'processes' resource, then guarantee no permissions, done check
+                    if not res["children"]:
+                        return response
+                    proc_res = res
+                    break
+            if not proc_res:
+                return response  # 'processes' itself does not exist, no permissions possible and done check
+
+            allowed_processes = []
+            known_processes = proc_res["children"].values()
+            known_processes = {res["node"].resource_name: res for res in known_processes}
+            request_user = get_user(response.request)
+            for proc_name in processes:
+                if proc_name not in known_processes:
+                    continue  # do not bother checking missing resource
+                child_proc = known_processes[proc_name]["node"]
+                perms = context.service.effective_permissions(request_user, child_proc, [Permission.READ])
+                if perms[0].access == Access.ALLOW:
+                    proc = processes[proc_name]
+                    allowed_processes.append(proc)
+
+            # override collected and permitted processes access by user
+            body["processes"] = allowed_processes
```

Comment:
Since the response doesn't contain any Weaver version information, I would at least make a basic check of the `processes` structure before overriding it in case that the format has evolved. If the structure doesn't fit, a warning could be raised so that we know that the hook doesn't support this version of Weaver anymore.

Reply:
I think it is more secure that way. If the structure evolved and processes were not overridden, this could make previously private processes to be leaked. The hook would need to be updated regardless to support the new structure, but processes would at least not be shown by mistake with current implementation.

Reply:
You're right!